### PR TITLE
feat(fetch): pin resolved addresses and add private-range and size policy

### PR DESCRIPTION
**Pull request by frostney**
## Summary

Closes the DNS-rebinding window in `fetch`, adds an opt-in private-range policy, and makes the response cap configurable. **WP-2** of the sandbox-hardening series. Stacked on #1130.

The allowlist was validated by **hostname string match**, and the connect that followed **re-resolved that hostname independently**. An allowlisted name under attacker DNS control could answer the check with a public address and the connect with an internal one — `169.254.169.254` being the usual target. The window reopened on every redirect hop.

- **Resolve once, validate, pin.** `DoRequest` now resolves the host a single time, validates the resolved address, and connects to that literal. Both platform connect paths already take a numeric fast path, so pinning adds no lookup and leaves no gap between check and use. Resolution, validation, and pinning are re-applied **per redirect hop**.
- **TLS still verifies against the hostname**, never the pinned literal. Pinning decides which address is dialed, not which identity the peer must prove; handing the literal to `StartTransportSecurity` would silently downgrade the connection's guarantees.
- **`--fetch-deny-private-ranges`** rejects targets resolving into RFC1918, loopback, link-local, CGNAT, or IPv6 ULA space, after resolution and on every hop. Opt-in — existing hosts see no change.
- **`--fetch-max-response-bytes`** makes the body ceiling configurable.

### Correction to the work package

The WP stated there is "no enforced response-body size cap visible in the request path." **There already was one** — an 8 MiB constant enforced across all three body paths (chunked, content-length, read-to-close), plus a 64 KiB header cap. It was *hard-coded*, not absent. This PR makes it configurable and keeps 8 MiB as the default, so behavior is unchanged unless a host opts in. The DNS-pinning and private-range halves of the WP stood exactly as written.

### Implementation constraints

Address classification is deliberately **deny-biased**: anything not exactly four decimal octets or a recognizable IPv6 form is treated as private rather than reinterpreted, because shortened and hex forms (`127.1`, `0x7f.0.0.1`) are a standard way to smuggle loopback past a textual filter.

Rejections raise `EHTTPError`, which the fetch manager already maps to a `TypeError` in script space, so all three surfaces match the existing "host not allowed" behavior in both execution modes.

`THTTPRequestPolicy` lives in `HTTPTypes`, not `HTTPClient`: the fetch manager's abstract base carries it in its interface and must not pull the socket implementation into its unit closure on the LAKON lane.

### Incidental fix

`scripts/fetch-e2e.sh` was **already failing on `main`** (1 passed / 10 failed) — every invocation lacked `--allowed-host`, so the run died before reaching any assertion. Fixed here because the new policy tests cannot run otherwise.

## Testing

- [x] Verified no regressions and confirmed the new feature or bugfix in end-to-end JavaScript/TypeScript tests
- [x] Updated documentation
- [x] Optional: Verified no regressions and confirmed the new feature or bugfix in native Pascal tests

`./build/GocciaTestRunner tests` **12143/12143** · `--mode=bytecode` **12143/12143** · `./format.pas --check` clean · `HTTPClient.Test` **10/10** (5 new: range classification, boundary cases, obfuscated-loopback rejection, literal round-trip, default-policy preservation) · `scripts/fetch-e2e.sh` **17/17** against a live server, including loopback-reachable-without-flag, loopback-refused-with-flag, redirect-hop-checked, and both cap directions.

Docs: `docs/built-ins-fetch.md` — options table plus a candid threat-model section (rebinding addressed; SSRF *reduced, not eliminated* — a public address proxying to an internal one is still out of reach).

### CODE REVIEW

**Review by frostney — COMMENTED** [review]
## Review notes (sandbox stack)

Resolve → validate → pin with TLS hostname preserved is the right SSRF shape. Policy host-global (not script-controllable), deny-biased IPv4 parsing, and the candid “reduced, not eliminated” docs are all good.

### Fix before undraft
1. **Size-cap hole on read-until-close** (`HTTPClient.pas` / `ReadResponse`). Body bytes already coalesced with headers (up to ~`MAX_RESPONSE_HEADER_BYTES`) are accepted without comparing to `AMaxBodyBytes`. Harmless at the old hard-coded 8 MiB; real with small `--fetch-max-response-bytes`. Enforce after `Copy(BodyBytes)` (and before exit). Content-Length / chunked look fine.
2. **“Redirect hop is policy-checked” e2e never leaves loopback.** `fetch-e2e.sh` hits `127.0.0.1/…/redirect` with deny-private on, so it fails on the *initial* resolve. Need public/allowlisted origin → private `Location`.

### Also worth doing
- **Canonicalize IPv4 literals** after `TryParseIPv4` to `%d.%d.%d.%d` before classify + dial. On Windows, `getaddrinfo` can treat leading-zero octets as octal (`0177.0.0.1` → classify public, connect loopback).
- Document that `--fetch-max-response-bytes=0` means “default 8 MiB”, or reject `0`.
- IPv6 classifier is incomplete (`fe80::/10`, multicast) — low urgency while connect is AF_INET-only, but a future footgun.

Opt-in private-range deny matches #1134’s ADR intent; hosts running untrusted script still need to turn it on.

Stack: sits on #1130; base for #1132.

**Comment by frostney** [comment]
@coderabbitai review

File: scripts/fetch-e2e.sh [hunk 1]
```diff
@@ -35,9 +35,57 @@ BASE="http://127.0.0.1:${PORT}"
...
+  "$LOADER" "$ALLOW_HOST" "$TMPFILE" --compat-asi 2>&1
+}
+
+# Same as run_js, but with extra loader flags before the script path, so the
+# policy flags can be exercised against the same live server.
+run_js_with() {
+  local extra="$1"
+  echo "$2" > "$TMPFILE"
+  # shellcheck disable=SC2086 -- $extra is a deliberate flag list
+  "$LOADER" "$ALLOW_HOST" $extra "$TMPFILE" --compat-asi 2>&1
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/fetch-e2e.sh:55]: Addressed in https://github.com/frostney/GocciaScript/commit/efdb52a6ab8fe63ad2df3c07fa8506e3fc822da3 (lands via #1139): Fixed the malformed ShellCheck directive.

File: scripts/fetch-e2e.sh [hunk 2]
```diff
@@ -150,6 +198,54 @@ check "POST is rejected" 1 \
...
+  console.log(e.message)
+}
+" "127.0.0.1"
+
+# A redirect hop must be resolved and validated exactly like the first
+# request; validating only the initial target would leave the hole open.
+check_with "redirect hop is policy-checked too" \
+  "--fetch-deny-private-ranges" 1 "
+await fetch('${BASE}/redirect')
+" "TypeError"
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/fetch-e2e.sh:234]: Addressed in https://github.com/frostney/GocciaScript/commit/efdb52a6ab8fe63ad2df3c07fa8506e3fc822da3 (lands via #1139): The redirect-hop test now exercises per-hop validation against a private target.

File: source/shared/HTTPClient.pas [hunk 3]
```diff
@@ -1077,27 +1094,205 @@ function ReadResponse(const ASock: TSocket;
...
+      Freeaddrinfo(Res);
+    end;
+  finally
+    Dispose(Hints);
+  end;
+  {$ENDIF}
+
+  if Result = '' then
+    raise EHTTPError.CreateFmt('Failed to resolve host: %s', [AHost]);
+end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/shared/HTTPClient.pas:1281]: Addressed in https://github.com/frostney/GocciaScript/commit/efdb52a6ab8fe63ad2df3c07fa8506e3fc822da3 (lands via #1139): Moved the socket-free URL helpers into HTTPTypes so the fetch builtin no longer pulls the socket unit (LAKON); HTTPClient re-exports, public API unchanged.
